File: fabric-network/chaincode/olivechain-domain/massbalance.go
// SPDX-License-Identifier: Apache-2.0
package main

import (
	"fmt"
	"math"

	"github.com/hyperledger/fabric-contract-api-go/v2/contractapi"
)

const defaultMassBalanceTolerance = 0.05

func (c *OliveChainContract) ReconcileResidue(ctx contractapi.TransactionContextInterface, residueID string) (*Reconciliation, error) {
	events, err := c.GetEventsForSubject(ctx, residueID, false)
	if err != nil {
		return nil, err
	}
	return reconcileFromEvents(ctx, residueID, events, defaultMassBalanceTolerance)
}

func reconcileFromEvents(ctx contractapi.TransactionContextInterface, residueID string, events []*Event, tolerance float64) (*Reconciliation, error) {
	rec := &Reconciliation{ResidueID: residueID, Outputs: map[string]float64{}, Issues: []string{}}
	for _, event := range events {
		effective, err := effectiveEvent(ctx, event)
		if err != nil {
			return nil, err
		}
		p := effective.Payload
		switch effective.EventType {
		case EventResidueGeneration:
			rec.GeneratedKg += asFloat(p["weight_kg"])
		case EventResidueCustody:
			rec.TransferredKg += asFloat(p["weight_kg"])
			if asBool(p["accepted"]) {
				rec.AcceptedKg += asFloat(p["weight_kg"])
			}
		case EventValorization:
			rec.TransformedKg += asFloat(p["input_kg"])
			rec.RejectedKg += asFloat(p["rejected_kg"])
			outputs, _ := p["outputs"].([]interface{})
			for _, raw := range outputs {
				output, _ := raw.(map[string]interface{})
				key := asString(output["output_type"])
				rec.Outputs[key] += asFloat(output["quantity"])
			}
		}
	}
	within := func(a, b float64) bool {
		max := math.Max(a, b)
		if max == 0 {
			return true
		}
		return math.Abs(a-b)/max <= tolerance
	}
	if rec.GeneratedKg == 0 {
		rec.Issues = append(rec.Issues, "no residue generation recorded")
	}
	if rec.TransferredKg != 0 && !within(rec.GeneratedKg, rec.TransferredKg) {
		rec.Issues = append(rec.Issues, fmt.Sprintf("transfer/generation mismatch: generated %vkg, transferred %vkg", rec.GeneratedKg, rec.TransferredKg))
	}
	if rec.AcceptedKg != 0 && !within(rec.TransferredKg, rec.AcceptedKg) {
		rec.Issues = append(rec.Issues, fmt.Sprintf("acceptance mismatch: transferred %vkg, accepted %vkg", rec.TransferredKg, rec.AcceptedKg))
	}
	accountedKg := rec.TransformedKg + rec.RejectedKg
	if accountedKg != 0 && !within(rec.AcceptedKg, accountedKg) {
		rec.Issues = append(rec.Issues, fmt.Sprintf(
			"transformation mismatch: accepted %vkg, transformed %vkg, rejected %vkg, accounted %vkg",
			rec.AcceptedKg,
			rec.TransformedKg,
			rec.RejectedKg,
			accountedKg,
		))
	}
	if accountedKg == 0 && rec.AcceptedKg > 0 {
		rec.Issues = append(rec.Issues, "residue accepted but no valorization recorded")
	}
	rec.Balanced = len(rec.Issues) == 0
	if rec.GeneratedKg > 0 {
		rec.RecoveryPct = round2(100 * rec.TransformedKg / rec.GeneratedKg)
	}
	return rec, nil
}

File: olivechain/massbalance.py
"""Mass-balance reconciliation for Loop 2.

Document: "This loop must preserve material balance. Rewards cannot be based
only on a declaration that waste was recycled. The system should reconcile
the quantity and characteristics of residues created, transported, accepted,
transformed, and converted into outputs."

Roadmap step 6: "Implement mass-balance and anomaly rules before enabling
rewards" — the incentives module refuses to issue rewards for a residue lot
until this reconciliation passes.
"""
from __future__ import annotations

from dataclasses import dataclass, field

from .ledger import Ledger
from .models import EventType


# Tolerance for measurement error and moisture loss between stages.
# The exact figure is a governance parameter to be agreed in pilot stage 2
# ("Specify the governance and data model"); 5% is a placeholder default.
DEFAULT_TOLERANCE = 0.05


@dataclass
class Reconciliation:
    residue_id: str
    generated_kg: float = 0.0
    transferred_kg: float = 0.0
    accepted_kg: float = 0.0
    transformed_kg: float = 0.0
    rejected_kg: float = 0.0
    outputs: dict = field(default_factory=dict)
    issues: list[str] = field(default_factory=list)

    @property
    def balanced(self) -> bool:
        return not self.issues

    @property
    def recovery_pct(self) -> float:
        if self.generated_kg <= 0:
            return 0.0
        return round(100.0 * self.transformed_kg / self.generated_kg, 2)


def reconcile_residue(ledger: Ledger, residue_id: str, tolerance: float = DEFAULT_TOLERANCE) -> Reconciliation:
    rec = Reconciliation(residue_id=residue_id)
    events = ledger.for_subject(residue_id)  # superseded events excluded

    for ev in events:
        p = ledger.effective_payload(ev)
        t = ev.event_type
        if t == EventType.RESIDUE_GENERATION.value:
            rec.generated_kg += float(p["weight_kg"])
        elif t == EventType.RESIDUE_CUSTODY.value:
            rec.transferred_kg += float(p["weight_kg"])
            if p.get("accepted"):
                rec.accepted_kg += float(p["weight_kg"])
        elif t == EventType.VALORIZATION.value:
            rec.transformed_kg += float(p["input_kg"])
            rec.rejected_kg += float(p.get("rejected_kg", 0))
            for o in p.get("outputs", []):
                key = o["output_type"]
                rec.outputs[key] = rec.outputs.get(key, 0) + float(o["quantity"])

    def within(a: float, b: float) -> bool:
        if max(a, b) == 0:
            return True
        return abs(a - b) / max(a, b) <= tolerance

    if rec.generated_kg == 0:
        rec.issues.append("no residue generation recorded")
    if rec.transferred_kg and not within(rec.generated_kg, rec.transferred_kg):
        rec.issues.append(
            f"transfer/generation mismatch: generated {rec.generated_kg}kg, transferred {rec.transferred_kg}kg"
        )
    if rec.accepted_kg and not within(rec.transferred_kg, rec.accepted_kg):
        rec.issues.append(
            f"acceptance mismatch: transferred {rec.transferred_kg}kg, accepted {rec.accepted_kg}kg"
        )
    accounted_kg = rec.transformed_kg + rec.rejected_kg
    if accounted_kg and not within(rec.accepted_kg, accounted_kg):
        rec.issues.append(
            "transformation mismatch: "
            f"accepted {rec.accepted_kg}kg, "
            f"transformed {rec.transformed_kg}kg, "
            f"rejected {rec.rejected_kg}kg, "
            f"accounted {accounted_kg}kg"
        )
    if accounted_kg == 0 and rec.accepted_kg > 0:
        rec.issues.append("residue accepted but no valorization recorded")

    return rec
